### PR TITLE
fix(traffic-split): upstream.type support chash configuration

### LINKED ISSUE #3858: bug: When using traffic-split ,  the key of chash is missing in the upstream of traffic-split
### Environment

* apisix version (2.3):
* OS (ubuntu 16):
* OpenResty / Nginx version (1.19.3):
* etcd version, if have (3.14):
* apisix-dashboard version, if have:

### Minimal test code / Steps to reproduce the issue

1.use traffic-split , and set the upstream as follow
```
curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/index.html",
    "plugins": {
        "traffic-split": {
            "rules": [
                {
                    "match": [
                        {
                            "vars": [
                                ["http_release","==","new_release"]
                            ]
                        }
                    ],
                    "weighted_upstreams": [
                        {
                            "upstream": {
                                "type": "chash",
                                "key": "remote_addr",
                                "name": "upstream_A",
                                "nodes": {
                                    "127.0.0.1:1981":1
                                    "127.0.0.2:1981":1
                                }
                            }
                        }
                    ]
                }
            ]
        }
    },
    "upstream": {
            "type": "roundrobin",
            "nodes": {
                "127.0.0.1:1980": 1
            }
    }
}'
```

2.
curl http://127.0.0.1:9080/index.html -H "release:new_release" -i
```
<html>
<head><title>500 Internal Server Error</title></head>
<body>
<center><h1>500 Internal Server Error</h1></center>
<hr><center>openresty</center>
</body>
</html>
```

### What's the actual result? (including assertion message & call stack if applicable)

Error can be found in error.log
```
2021/03/18 15:36:04 [error] 72662#72662: *683 failed to run balancer_by_lua*: /chash.lua:36: invalid argument, expect string value
stack traceback:
        [C]: in function 'error'
        /data/app/huya-nginx-proxy/apisix/core/ctx.lua:139: in function '__index'
        /data/app/huya-nginx-proxy/apisix/balancer/chash.lua:36: in function 'fetch_chash_hash_key'
        /data/app/huya-nginx-proxy/apisix/balancer/chash.lua:81: in function 'get'
        /data/app/huya-nginx-proxy/apisix/balancer.lua:172: in function 'pick_server'
        /data/app/huya-nginx-proxy/apisix/balancer.lua:197: in function 'load_balancer'
        /data/app/huya-nginx-proxy/apisix/init.lua:718: in function 'http_balancer_phase'
```

### What's the expected result?
No error

### PR DESCRIPTION
**Pull request by Firstsawyou**
fix #3858
Signed-off-by: Firstsawyou yuelinz99@gmail.com

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: docs/zh/latest/plugins/traffic-split.md [hunk 1]
```diff
@@ -47,7 +47,7 @@ traffic-split 插件使用户可以逐步引导各个上游之间的流量百分
 | rules.weighted_upstreams       | array[object] | 可选   |        |        | 上游配置规则列表。 |
 | weighted_upstreams.upstream_id | string / integer | 可选   |        |        | 通过上游 id 绑定对应上游。 |
 | weighted_upstreams.upstream    | object | 可选   |        |        | 上游配置信息。 |
-| upstream.type                  | enum   | 可选   |   roundrobin |  [roundrobin, chash]      | roundrobin 支持权重的负载，chash 一致性哈希，两者是二选一的(目前只支持 `roundrobin`)。 |
+| upstream.type                  | enum   | 可选   |   roundrobin |  [roundrobin, chash]      | roundrobin 支持权重的负载，chash 一致性哈希，两者是二选一。 |
```
[the remark this answers is not in the document]
Comment (tokers) [reply docs/zh/latest/plugins/traffic-split.md:50]: I think it's necessary. @Firstsawyou

[the remark this answers is not in the document]
Comment (Firstsawyou) [reply docs/zh/latest/plugins/traffic-split.md:50]: Yes, agree.

File: t/plugin/traffic-split2.t [hunk 2]
```diff
@@ -311,3 +311,93 @@ host: localhost
...
+        end
+
+        table.sort(bodys)
+        ngx.say(table.concat(bodys, ", "))
+    }
+}
+--- request
+GET /t
+--- response_body eval
+qr/1981, 1981, 1981, 1981|1982, 1982, 1982, 1982/
```
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply t/plugin/traffic-split2.t:392]: I think we can add another header to trigger the use of 1982.

[the remark this answers is not in the document]
Comment (Firstsawyou) [reply t/plugin/traffic-split2.t:392]: I think we can add another header to trigger the use of 1982 ？

[the remark this answers is not in the document]
Comment (spacewander) [reply t/plugin/traffic-split2.t:392]: Agree

File: docs/zh/latest/plugins/traffic-split.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply docs/zh/latest/plugins/traffic-split.md:50]: Updated.

File: t/plugin/traffic-split2.t [hunk 3]
```diff
@@ -311,3 +311,93 @@ host: localhost
...
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
```
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply t/plugin/traffic-split2.t:365]: Updated.

File: apisix/plugins/traffic-split.lua [hunk 4]
```diff
@@ -200,6 +200,11 @@ local function set_upstream(upstream_info, ctx)
         }
     }
 
+    if upstream_info.type == "chash" then
+        up_conf.hash_on = upstream_info.hash_on
+        up_conf.key = upstream_info.key
```
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply apisix/plugins/traffic-split.lua]: updated.

File: t/plugin/traffic-split2.t [hunk 5]
```diff
@@ -311,3 +311,93 @@ host: localhost
...
+                    }
+                }]=]
+            )
+            if code >= 300 then
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+--- request
```
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply t/plugin/traffic-split2.t]: updated.

File: t/plugin/traffic-split2.t [hunk 6]
```diff
@@ -376,28 +374,38 @@ location /t {
         local t = require("lib.test_admin").test
         local bodys = {}
         local headers = {}
+        local headers2 = {}
         headers["custom_header"] = "custom-one"
-        for i = 1, 4 do
+        headers2["custom_header"] = "customtow"
```
Comment (spacewander) [inline t/plugin/traffic-split2.t:379]: Better to avoid using a word that looks like typo

File: t/plugin/traffic-split2.t [hunk 7]
```diff
@@ -311,3 +311,101 @@ host: localhost
...
+            bodys[i] = body
+            bodys[i+1] = body2
+        end
+
+        table.sort(bodys)
+        ngx.say(table.concat(bodys, ", "))
+    }
+}
+--- response_body eval
+qr/1981, 1981, 1981, 1981, 1982, 1982, 1982, 1982/
```
Comment (spacewander) [inline t/plugin/traffic-split2.t:392]: Should be 1981, 1982, ...?

File: t/plugin/traffic-split2.t [hunk 8]
```diff
@@ -311,3 +311,101 @@ host: localhost
...
+        headers["custom_header"] = "custom-one"
+        headers2["custom_header"] = "customtow"
+        for i = 1, 8, 2 do
+            local _, _, body = t('/server_port', ngx.HTTP_GET, "", nil, headers)
+            local _, _, body2 = t('/server_port', ngx.HTTP_GET, "", nil, headers2)
+            bodys[i] = body
+            bodys[i+1] = body2
+        end
+
+        table.sort(bodys)
```
Comment (spacewander) [inline t/plugin/traffic-split2.t:387]: We should remove this line as it breaks the order introduced by chash.

**Review by Firstsawyou — COMMENTED** [review]
@spacewander  Updated.

File: t/plugin/traffic-split2.t (hunk 8 above)
Comment (Firstsawyou) [reply t/plugin/traffic-split2.t]: Updated.

File: t/plugin/traffic-split2.t (hunk 6 above)
Comment (Firstsawyou) [reply t/plugin/traffic-split2.t]: Updated.

**Review by spacewander — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Pull request merged**